datumaro/util/scope.py

```diff
@@ -0,0 +1,171 @@
+                except: # pylint: disable=bare-except
+                    if not self.ignore_errors:
+                        raise
+
+    def __init__(self):
+        self._handlers = {}
+        self._error_stack = ExitStack()
+        self._exit_stack = ExitStack()
+        self.enabled = True
+
+    def on_error_do(self, callback, *args, name: Optional[str] = None,
+            enabled: bool = True, ignore_errors: bool = False,
+            fwd_kwargs: Optional[Dict[str, Any]] = None, **kwargs) -> str:
+        """
+        Registers a function to be called on scope exit because of an error.
+        Equivalent to the "except" block of "try-except".
+        """
+
+        if args or kwargs or fwd_kwargs:
+            if fwd_kwargs:
+                kwargs.update(fwd_kwargs)
+            callback = partial(callback, *args, **kwargs)
+
+        name = name or hash(callback)
+        assert name not in self._handlers, "Callback is already registered"
+
+        handler = self.Handler(callback,
+            enabled=enabled, ignore_errors=ignore_errors)
+        self._handlers[name] = handler
+        self._error_stack.callback(handler)
+        return name
+
+    def on_exit_do(self, callback, *args, name: Optional[str] = None,
+            enabled: bool = True, ignore_errors: bool = False,
+            fwd_kwargs: Optional[Dict[str, Any]] = None, **kwargs) -> str:
+        """
+        Registers a function to be called on scope exit unconditionally.
+        Equivalent to the "finally" block of "try-except".
+        """
+
+        if args or kwargs or fwd_kwargs:
+            if fwd_kwargs:
+                kwargs.update(fwd_kwargs)
+            callback = partial(callback, *args, **kwargs)
+
+        name = name or hash(callback)
+        assert name not in self._handlers, "Callback is already registered"
+
+        handler = self.Handler(callback,
+            enabled=enabled, ignore_errors=ignore_errors)
+        self._handlers[name] = handler
+        self._exit_stack.callback(handler)
+        return name
+
+    def add(self, cm: ContextManager) -> Any:
+        """
+        Enters a context manager and adds it to the exit stack.
+        """
+
+        return self._exit_stack.enter_context(cm)
+
+    def enable(self, name=None):
+        if name:
+            self._handlers[name].enabled = True
+        else:
+            self.enabled = True
+
+    def disable(self, name=None):
+        if name:
+            self._handlers[name].enabled = False
+        else:
+            self.enabled = False
+
+    def clean(self):
+        self.__exit__()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type=None, exc_value=None, exc_traceback=None):
+        if not self.enabled:
+            return
+
+        try:
+            if exc_type:
+                self._error_stack.__exit__(exc_type, exc_value, exc_traceback)
+        finally:
+            self._exit_stack.__exit__(exc_type, exc_value, exc_traceback)
```

Comment:
So the idea is to wrap each callback in a function and check the exception type inside?

Reply:
You're already wrapping each callback in a `Handler` object - you just need to use `push` instead of `callback` and adjust the arguments of `Handler.__call__` accordingly.

Reply:
Done